sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java

```diff
@@ -305,27 +324,53 @@ public void startBundle(StartBundleContext c) throws Exception {
     }
 
     @ProcessElement
+    @SuppressWarnings("ReferenceEquality")
     public void processElement(ProcessContext c) throws Exception {
-      List<OutgoingMessage> pubsubMessages = new ArrayList<>(publishBatchSize);
-      int bytes = 0;
+      // TODO(sjvanrossum): Refactor the write transform so this map can be indexed with topic +
+      // ordering key and have bundle scoped lifetime.
+      // NB: A larger, breaking refactor could make this irrelevant with a GBK on topic + ordering
+      // key (or GroupIntoBatches with configurable shard for ShardedKey?) and unify
+      // bounded/unbounded writes and static/dynamic destinations.
```

Comment:
Yeah, I've just put together an experiment for it and this seems worth it.

Reply:
Deferring to apache/beam#31828 since batching with GIB/GIB-WSK depends on apache/beam#31800, which turned out to be much broader in scope than expected (~500 lines patch to fix it).

Reply:
Thanks for creating those issues. I like the idea of deferring this to a `#withAutoSharding()` method